### PR TITLE
FINERACT-1806: Trigger event when charge-off happens

### PR DESCRIPTION
**Pull request by adamsaghy**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+    "name": "UnpaidChargeDataV1",
+    "namespace": "org.apache.fineract.avro.loan.v1",
+    "type": "record",
+    "fields": [
+        {
+            "default": null,
+            "name": "chargeId",
+            "type": [
+                "null",
```
Comment (galovics) [inline fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc]: This cannot be null, can it?

File: fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+                "long"
+            ]
+        },
+        {
+            "default": null,
+            "name": "chargeName",
```
Comment (galovics) [inline fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc:11]: Same here, name is mandatory for charges, isn't it?

File: fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            "type": [
+                "null",
+                "string"
+            ]
+        },
+        {
+            "default": null,
+            "name": "outstandingAmount",
```
Comment (galovics) [inline fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc:15]: Can this be null or only zero?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/data/UnpaidChargeData.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * Immutable data object representing the total unpaid charge details
+ */
+@Getter
+public class UnpaidChargeData {
+
+    private final Long chargeId;
+    private final String chargeName;
+    private final BigDecimal outstandingAmount;
+
+    public UnpaidChargeData(Long chargeId, String chargeName, BigDecimal outstandingAmount) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/data/UnpaidChargeData.java:36]: Note if what I wrote in the Avro schema is true (i.e. not null values) then let's add null checks here to ensure we cannot create this object without having proper data in it.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/event/external/service/serialization/mapper/loan/LoanTransactionDataMapper.java [hunk 5]
```diff
@@ -22,9 +22,11 @@
 import org.apache.fineract.infrastructure.event.external.service.serialization.mapper.support.AvroMapperConfig;
 import org.apache.fineract.portfolio.loanaccount.data.LoanTransactionData;
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
 
 @Mapper(config = AvroMapperConfig.class)
 public interface LoanTransactionDataMapper {
 
+    @Mapping(target = "unpaidCharges", ignore = true)
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/event/external/service/serialization/mapper/loan/LoanTransactionDataMapper.java:31]: Can we put a comment here why this is ignored? Just for future reference.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/event/external/service/serialization/mapper/loan/LoanAccountDataMapper.java [hunk 6]
```diff
@@ -36,5 +36,6 @@ public interface LoanAccountDataMapper {
     @Mapping(source = "interestRecalculationEnabled", target = "isInterestRecalculationEnabled")
     LoanAccountDataV1 map(LoanAccountData source);
 
+    @Mapping(target = "unpaidCharges", ignore = true)
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/event/external/service/serialization/mapper/loan/LoanAccountDataMapper.java]: I don't think we even need this method here.
We could just modify the annotation on the class to this:
```
@Mapper(config = AvroMapperConfig.class, uses = { LoanTransactionDataMapper.class })
```

**Review by galovics — CHANGES_REQUESTED** [review]

File: fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc (hunk 1 above)
Comment (adamsaghy) [reply fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc]: Fair point!

File: fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc (hunk 3 above)
Comment (adamsaghy) [reply fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc:15]: Greater than 0

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/event/external/service/serialization/mapper/loan/LoanAccountDataMapper.java (hunk 6 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/event/external/service/serialization/mapper/loan/LoanAccountDataMapper.java]: I give a try :)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/data/UnpaidChargeData.java (hunk 4 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/data/UnpaidChargeData.java:36]: sure

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/event/external/service/serialization/mapper/loan/LoanTransactionDataMapper.java (hunk 5 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/event/external/service/serialization/mapper/loan/LoanTransactionDataMapper.java:31]: sure

File: fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc (hunk 2 above)
Comment (adamsaghy) [reply fineract-avro-schemas/src/main/avro/loan/v1/UnpaidChargeDataV1.avsc:11]: yes

**Review by galovics — APPROVED** [review]

**Pull request merged**